test/graph_compaction_tests.cc

```diff
@@ -65,78 +66,51 @@ namespace detail {
-			CHECK(buf_a_last_writer_map_size() <= NUM_NODES * 2);
-			for(node_id n = 0; n < NUM_NODES; ++n) {
-				CHECK(inspector.get_commands(std::nullopt, n, command_type::HORIZON).size() <= NUM_TIMESTEPS);
-				CHECK(inspector.get_commands(std::nullopt, n, command_type::HORIZON).size() >= NUM_TIMESTEPS - 1);
-			}
+			const auto latest_write_accessor = [t = timestep](celerity::chunk<1> chnk) {
+				celerity::subrange<2> ret;
+				ret.range = cl::sycl::range<2>(1, chnk.range.size());
+				ret.offset = cl::sycl::id<2>(t, chnk.offset.get(0));
+				return ret;
+			};
 
-			// also check that unused commands are deleted
-			CHECK(ctx.get_command_graph().command_count() <= NUM_NODES * 13);
-			// and are removed from the read cache
-			CHECK(graph_generator_testspy::get_command_buffer_reads_size(ctx.get_graph_generator()) < NUM_NODES * 9);
+			test_utils::build_and_flush(ctx, NUM_NODES,
+			    test_utils::add_compute_task<class growing_read_kernel>(
+			        ctx.get_task_manager(),
+			        [&](handler& cgh) {
+				        buf_a.get_access<mode::read>(cgh, read_accessor);
+				        buf_a.get_access<mode::discard_write>(cgh, latest_write_accessor);
+			        },
+			        full_range));
 		}
 
-		SECTION("with horizon step size 3") {
-			ctx.get_task_manager().set_horizon_step(3);
-
-			SECTION("and a growing read pattern") { time_series_lambda(true); }
-			SECTION("and a latest-only read pattern") { time_series_lambda(false); }
-
-			CHECK(buf_a_region_map_size() <= NUM_NODES * 2 * 3);
-			CHECK(buf_a_last_writer_map_size() <= NUM_NODES * 2 * 3);
-			for(node_id n = 0; n < NUM_NODES; ++n) {
-				CHECK(inspector.get_commands(std::nullopt, n, command_type::HORIZON).size() <= NUM_TIMESTEPS / 3 + 1);
-				CHECK(inspector.get_commands(std::nullopt, n, command_type::HORIZON).size() >= NUM_TIMESTEPS / 3 - 1);
-			}
-
-			// also check that unused commands are deleted
-			CHECK(ctx.get_command_graph().command_count() <= NUM_NODES * 13 * 3);
-			// and are removed from the read cache
-			CHECK(graph_generator_testspy::get_command_buffer_reads_size(ctx.get_graph_generator()) < NUM_NODES * 9 * 3);
+		const auto buf_a_region_map_size = graph_generator_testspy::get_buffer_states_num_regions(ctx.get_graph_generator(), buf_a.get_id());
+		CHECK(buf_a_region_map_size <= NUM_NODES * 2 * horizon_step_size);
+		const auto buf_a_last_writer_map_size = graph_generator_testspy::get_buffer_last_writer_num_regions(ctx.get_graph_generator(), buf_a.get_id());
+		CHECK(buf_a_last_writer_map_size <= NUM_NODES * 2 * horizon_step_size);
+		for(node_id n = 0; n < NUM_NODES; ++n) {
+			CHECK(inspector.get_commands(std::nullopt, n, command_type::HORIZON).size() <= NUM_TIMESTEPS / horizon_step_size + 1);
+			CHECK(inspector.get_commands(std::nullopt, n, command_type::HORIZON).size() >= NUM_TIMESTEPS / horizon_step_size - 1);
 		}
 
+		// also check that unused commands are deleted
+		CHECK(ctx.get_command_graph().command_count() <= NUM_NODES * 13 * horizon_step_size);
```

Comment:
Do we now know where that 13 (and 6 below) is coming from? If so consider adding a comment.

Reply:
I have now reverse engineered and documented those values.

Reply:
I have now reverse engineered and documented these bounds.

Reply:
Wow, looks like this took a lot of work!